### ISSUE TITLE
Create a TableOfContents module

### ISSUE THREAD

**Issue by daveaglick** [statiqdev/Statiq#29]
Automatically extracts headings from a HTML document and generates a table of contents. How to specify where the ToC is inserted? How to identify whether a heading has an anchor (and what to do if it doesn't)? Should it also be added to the metadata? What if we just want it added to metadata without inserting the ToC (I.e., for display on another page)?

Possibly call this modules Headings instead of ToC (or maybe two different complimentary modules).

**Comment by avishnyakov** [comment]
ToC generation would be something interesting to see yet hard to implement in a generic way.
Maybe it would be reasonable to separate data and UI?

For instance a module to generate a ToC tree data structure based on folders, metadata or combination of folders + medata would be nice to have. Right now the documents collection is flat, so it takes some effort to generate a tree structure based on docs collection. Even once a tree structure generated it is challenging to render due to lack of helper and function support (mean Razor templates). That's the experience so far.

Having a module that would be able to process documents and push a tree structure into metedata flow would be a nice start. Once that done then the rendering a ToC data tree would not be a problem.

**Comment by daveaglick** [comment]
> Having a module that would be able to process documents and push a tree structure into metedata flow would be a nice start

This got me thinking - this is something that's been lacking. I created the `ToLookup()` methods to try and address the need for grouping documents which indirectly addresses this a little bit, but that's not far enough. Check out #150 to see what I came up with (comments welcome!).

**Comment by FlorianRappl** [comment]
I will implement this for HTML - it will be split in two modules. One generates a TOC (from HTML) to a TOC model and places it in a metadata key (which can be specified if needed), the other one transforms the TOC model to HTML from a given (or the default) metadata key.

The idea is then to use it in combination with, e.g., `Replace`. Also while the transformer (from the TOC model) will be designed flexible, it may not fit everyone's purpose. Hence the reason for splitting the whole module into two.

What do you think about this?

Should I make it a core module (specifically: two core modules), or provide it via NuGet?

**Comment by daveaglick** [comment]
@FlorianRappl Great!

I think the two-module approach as you've proposed is the way to go. I can envision all sorts of uses for the set of headings besides inserting a ToC right on the page. For example, aggregating all the ToC info from a set of pages for insertion in some kind of index.

The part that keeps bugging me is what to do about headings that don't already have anchors. Perhaps an option to insert missing anchors when running the module that populates the metadata?

I try to keep dependencies for the core library minimal, so these should probably go in Wyam.Modules.Html. While they could go into a totally separate library, that one already has AngleSharp and I think it makes sense to consolidate when possible.

**Comment by FlorianRappl** [comment]
Yes, the `Modules.Html` seems to be most appropriate. Also AngleSharp can be used to insert these named anchor tags, however, I would distinguish between just inserting a TOC and actually pinning TOC entries to sections. While the standard TOC transformer may do such a thing (and would therefore also inset named anchors if not available), the TOC extractor does nothing like it.

Therefore I propose the following. I will make a PR to the `Modules.Html` library with a module proposal "TOC extractor". Once this one is done (this includes a simple TOC model) we can discuss the design / implementation of the standard "TOC transformer". But let's start with the simpler / more basic component first.

**Comment by daveaglick** [comment]
:+1: sounds good

**Comment by FlorianRappl** [comment]
I know its been over 2 months; but I now have more time and since its still open I will try to take care of it. However, the transition to markdig is more important imho. I will prioritize accordingly.

**Comment by daveaglick** [comment]
@FlorianRappl No problem! I can completely relate to the ups and downs of availability. Also agree with your priority assessment. If you don't get to this one in this cycle, we'll get there eventually... :smile:

**Comment by daveaglick** [comment]
This is now implemented in a `Headings` module in the develop branch. It queries for heading elements at a user-specified depth and then adds them in the metadata of the input document as a list of documents (nested or not, configurable by user).
